pkg/didcomm/transport/http/outbound_test.go

```diff
@@ -0,0 +1,183 @@
+	"testing"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/require"
+)
+
+const certPrefix = "../../../../test/fixtures/keys/"
+
+func TestWithOutboundOpts(t *testing.T) {
+	opt := WithOutboundCommHTTP(nil)
+	require.NotNil(t, opt)
+	clOpts := &outboundCommHTTPOpts{}
+	opt(clOpts)
+
+	opt = WithOutboundTimeout(10 * time.Second)
+	require.NotNil(t, opt)
+	clOpts = &outboundCommHTTPOpts{}
+	// opt.client is nil, so setting timeout should panic
+	require.Panics(t, func() { opt(clOpts) })
+
+	opt = WithOutboundTLSConfig(nil)
+	require.NotNil(t, opt)
+	clOpts = &outboundCommHTTPOpts{}
+	opt(clOpts)
+}
+
+func TestOutboundHTTPTransport(t *testing.T) {
+	// prepare http server
+	server := startMockServer()
+	defer func() {
+		err := server.Close()
+		if err != nil {
+			log.Fatalf("Failed to stop server: %s", err)
+		}
+	}()
+	//build a mock cert pool
+	cp := x509.NewCertPool()
+	err := addCertsToCertPool(cp)
+	require.NoError(t, err)
+
+	// build a tls.Config instance to be used by the outbound transport
+	tlsConfig := &tls.Config{
+		RootCAs:      cp,
+		Certificates: nil,
+	}
+	// create a new invalid Outbound transport instance
+	ot, err := NewOutbound()
+	require.Error(t, err)
+	require.EqualError(t, err, "Can't create an outbound transport without an HTTP client or tls.Config option")
+
+	// now create a new valid Outbound transport instance and test its Send() call
+	ot, err = NewOutbound(WithOutboundTLSConfig(tlsConfig), WithOutboundTimeout(1*time.Second))
+	require.NoError(t, err)
+	require.NotNil(t, ot)
+
+	// test mock HTTPHandler
+	req, err := http.NewRequest("POST", "https://localhost:8090", nil)
+	require.NoError(t, err)
+	req.Header.Set("Content-type", commContentType)
+
+	rr := httptest.NewRecorder()
+	server.Handler.ServeHTTP(rr, req)
+
+	// finally test Outbound transport's api
+	// first with an empty url
+	r, e := ot.Send("Hello World", "")
+	require.Error(t, e)
+	require.Empty(t, r)
+
+	// now try a bad url
+	r, e = ot.Send("Hello World", "https://badurl")
+	require.Error(t, e)
+	require.Empty(t, r)
+
+	// and try with a 'bad' payload with a valid url..
+	r, e = ot.Send("bad", "https://localhost:8090")
+	require.Error(t, e)
+	require.Empty(t, r)
+
+	// finally using a valid url
+	r, e = ot.Send("Hello World", "https://localhost:8090")
+	require.NoError(t, e)
+	require.NotEmpty(t, r)
+
+}
+
+func addCertsToCertPool(pool *x509.CertPool) error {
+	var rawCerts []string
+
+	for i := 1; i <= 3; i++ {
+		certPath := fmt.Sprintf("%sec-pubCert%d.pem", certPrefix, i)
+		// Create a pool with server certificates
+		cert, e := ioutil.ReadFile(filepath.Clean(certPath))
+		if e != nil {
+			return errors.Wrap(e, "Failed Reading certificate")
+		}
+		rawCerts = append(rawCerts, string(cert))
+	}
+
+	certs := decodeCerts(rawCerts)
+	for i := range certs {
+		pool.AddCert(certs[i])
+	}
+	return nil
+}
+
+func startMockServer() *http.Server {
+	testHandler := mockHttpHandler{}
+	httpServer := &http.Server{
+		Addr:    ":8090",
```

Comment:
This can fail if the port is already used. You should have the http.Server supply the port back and then use that returned port in the tests.

Reply:
updated to get dynamic port with net.Listener ..